File: Behavioral/iterator/examples/Collection/Python/Collection/GenericColletion/GenericEnumerator.py
from GenericColletion.GenericEnumeratorBase import GenericEnumeratorBase


class GenericEnumerator(GenericEnumeratorBase):
    _readonly_collection = []
    _current_index = -1

    def __init__(self, readonly_collection: list):
        self._readonly_collection = readonly_collection

    def current(self) -> object:
        if self._current_index < 0:
            raise TypeError("Call move_next() first!")
        return self._readonly_collection[self._current_index]

    def move_next(self) -> bool:
        if self._current_index == len(self._readonly_collection) - 1:
            return False

        self._current_index += 1
        return True

    def reset(self):
        self._current_index = -1

File: Behavioral/iterator/examples/Collection/C#/Collection/Collection/GenericCollection/GenericEnumerator.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace Collection.GenericCollection
{
    class GenericEnumerator : IGenericEnumerator
    {
        private int _currentIndex { get; set; }
        private object[] _readOnlyCollection { get; }

        public object Current
        {
            get
            {
                if (_currentIndex < 0)
                    throw new InvalidOperationException("Call MoveNext() first!");

                return _readOnlyCollection[_currentIndex];
            }

        }


        public GenericEnumerator(object[] readOnlyCollection)
        {
            _readOnlyCollection = readOnlyCollection;
            _currentIndex = -1;
        }

        public bool MoveNext()
        {
            if (_currentIndex == _readOnlyCollection.Length-1)
                return false;

            _currentIndex++;
            return true;
        }

        public void Reset()
        {
            _currentIndex = 0;
        }
    }
}
